<a href="https://colab.research.google.com/github/saurabhdaund76/Book-Recommendation-System/blob/main/Book_recommendation_system.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [2]:
# reading user csv
user_df = pd.read_csv('/content/drive/MyDrive/ML projects/Users.csv')

In [3]:
# reading ratings csv
ratings_df = pd.read_csv('/content/drive/MyDrive/ML projects/Ratings.csv')

In [27]:
# reading books csv
books_df = pd.read_csv('/content/drive/MyDrive/ML projects/Books.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [10]:
# shape of all df

print(ratings_df.shape)
print(user_df.shape)
print(books_df.shape)

(1149780, 3)
(278858, 3)
(271360, 8)


In [6]:
ratings_df.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [11]:
ratings_df['User-ID'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
158698        1
17920         1
277135        1
275086        1
187812        1
Name: User-ID, Length: 105283, dtype: int64

In [23]:
ratings_df.groupby('User-ID')['Book-Rating'].sum().reset_index()

User-ID  Book-Rating
0             2            0
1             7            0
2             8           39
3             9            6
4            10            6
...         ...          ...
105278   278846            8
105279   278849            9
105280   278851           91
105281   278852            8
105282   278854           42

[105283 rows x 2 columns]

In [28]:
books_df.head()

ISBN  ...                                        Image-URL-L
0  0195153448  ...  http://images.amazon.com/images/P/0195153448.0...
1  0002005018  ...  http://images.amazon.com/images/P/0002005018.0...
2  0060973129  ...  http://images.amazon.com/images/P/0060973129.0...
3  0374157065  ...  http://images.amazon.com/images/P/0374157065.0...
4  0393045218  ...  http://images.amazon.com/images/P/0393045218.0...

[5 rows x 8 columns]

so here we can see we have some extra columns which are not required like image url s m l so we will drop them

In [31]:
books_df = books_df.drop(['Image-URL-S','Image-URL-M','Image-URL-L'], axis=1)

In [32]:
books_df.head()

ISBN  ...                   Publisher
0  0195153448  ...     Oxford University Press
1  0002005018  ...       HarperFlamingo Canada
2  0060973129  ...             HarperPerennial
3  0374157065  ...        Farrar Straus Giroux
4  0393045218  ...  W. W. Norton &amp; Company

[5 rows x 5 columns]

Data preprocessing 

In [33]:
books_df.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
dtype: int64

In [34]:
user_df.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [35]:
ratings_df.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

as we can see we only have null values in books df and user df

In [37]:
# we deleted the cells with empty rows
books_df = books_df.dropna(axis=0)

In [38]:
books_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 271357 entries, 0 to 271359
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271357 non-null  object
 1   Book-Title           271357 non-null  object
 2   Book-Author          271357 non-null  object
 3   Year-Of-Publication  271357 non-null  object
 4   Publisher            271357 non-null  object
dtypes: object(5)
memory usage: 12.4+ MB


as we can see there are no null values

As we can see year of publication is object type it should be numeric type

In [44]:
books_df['Year-Of-Publication'].astype

<bound method NDFrame.astype of 0         2002
1         2001
2         1991
3         1999
4         1999
          ... 
271355    1988
271356    1991
271357    2004
271358    1996
271359    2000
Name: Year-Of-Publication, Length: 271357, dtype: object>

In [48]:
books_df['Year-Of-Publication'].unique()


array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

Here we have some of the string values as well as we have invalid years ie years above 2021 so we will try nad handle these values

In [58]:
print(len(books_df[books_df['Year-Of-Publication']=='DK Publishing Inc']))
print(len(books_df[books_df['Year-Of-Publication']=='Gallimard']))
print(len(books_df[books_df['Year-Of-Publication']==0]))

2
1
3570


As we can see there are around 3570 values which have 0 as a publication year

In [62]:
books_df.describe()

ISBN      Book-Title  ... Year-Of-Publication  Publisher
count       271357          271357  ...              271357     271357
unique      271357          242133  ...                 202      16806
top     0140504230  Selected Poems  ...                2002  Harlequin
freq             1              27  ...               13902       7535

[4 rows x 5 columns]

as we can see top value in publication is 2002 so we will replace all these values with 2002 including the year greater then 2021


In [63]:
books_df[books_df['Year-Of-Publication']=='DK Publishing Inc'] = 2002
books_df[books_df['Year-Of-Publication']=='Gallimard']=  2002

In [64]:
books_df['Year-Of-Publication'].unique()
# we replaced the string values by year 

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

Now we will convert object type to int type for publication column

In [68]:
books_df['Year-Of-Publication'] = books_df['Year-Of-Publication'].astype(int)


Now we will change all 0 years with max year and all the years above 2021 by max year ie 2002

In [70]:
books_df.loc[books_df['Year-Of-Publication'] > 2021, 'Year-Of-Publication'] = 2002
books_df.loc[books_df['Year-Of-Publication'] == 0, 'Year-Of-Publication'] = 2002

In [72]:
# dropping duplicates
books_df.drop_duplicates(keep='last', inplace=True) 
books_df.reset_index(drop = True, inplace = True)

Users data preprocessing

In [74]:
user_df.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [78]:
# as we have seen we have missing values in age column 
user_df['Age'].unique()

array([ 34.7514337,  18.       ,  17.       ,  61.       ,  26.       ,
        14.       ,  25.       ,  19.       ,  46.       ,  55.       ,
        32.       ,  24.       ,  20.       ,  34.       ,  23.       ,
        51.       ,  31.       ,  21.       ,  44.       ,  30.       ,
        57.       ,  43.       ,  37.       ,  41.       ,  54.       ,
        42.       ,  50.       ,  39.       ,  53.       ,  47.       ,
        36.       ,  28.       ,  35.       ,  13.       ,  58.       ,
        49.       ,  38.       ,  45.       ,  62.       ,  63.       ,
        27.       ,  33.       ,  29.       ,  66.       ,  40.       ,
        15.       ,  60.       ,   0.       ,  79.       ,  22.       ,
        16.       ,  65.       ,  59.       ,  48.       ,  72.       ,
        56.       ,  67.       ,   1.       ,  80.       ,  52.       ,
        69.       ,  71.       ,  73.       ,  78.       ,   9.       ,
        64.       , 103.       , 104.       ,  12.       ,  74. 

In [76]:
# fill missing values with mean
user_df['Age'].fillna(user_df.Age.mean(), inplace=True)


In [77]:
user_df['Age'].isnull().sum()

0

here age values are too high which even practically seems impossible so we will take only age greater then 5 and less then 80 sp we will substitute all these values with mean


In [82]:
user_df.loc[user_df['Age']>=80, 'Age'] = round(user_df['Age'].mean())
user_df.loc[user_df['Age']<=10, 'Age'] = round(user_df['Age'].mean())

In [83]:
user_df['Age'] = user_df['Age'].astype(int)       #changing Datatype to int


In [87]:
## Drop duplicate rows
user_df.drop_duplicates(keep='last', inplace=True)
user_df.reset_index(drop=True, inplace=True)

In [90]:
user_df.head()

User-ID                            Location  Age
0        1                  nyc, new york, usa   34
1        2           stockton, california, usa   18
2        3     moscow, yukon territory, russia   34
3        4           porto, v.n.gaia, portugal   17
4        5  farnborough, hants, united kingdom   34

In [91]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   User-ID   278858 non-null  int64 
 1   Location  278858 non-null  object
 2   Age       278858 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 6.4+ MB


ratings data pre processing

In [92]:
ratings_df.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [93]:
ratings_df['ISBN'].unique()

array(['034545104X', '0155061224', '0446520802', ..., '0679752714',
       '0806917695', '05162443314'], dtype=object)